File: packages/rslint/src/config/base-path.ts
import path from 'node:path';

function isRecord(value: unknown): value is Record<string, unknown> {
  return value !== null && typeof value === 'object' && !Array.isArray(value);
}

/** True for absolute POSIX or Windows paths (including drive-letter forms). */
function isAbsolutePattern(pattern: string): boolean {
  return path.posix.isAbsolute(pattern) || path.win32.isAbsolute(pattern);
}

/**
 * Escape glob metacharacters in a resolved basePath so the directory is treated
 * literally when spliced into glob patterns. ESLint treats basePath as a literal
 * path; without escaping, basePath "packages/[locale]" would be interpreted as a
 * character class and never match. Character-class escapes ([*], [?], [[] ...)
 * are understood by both the Go (doublestar) and TS (picomatch) matchers on
 * every platform.
 */
function escapeGlobBasePath(s: string): string {
  return s
    .replaceAll('[', '[[]')
    .replaceAll('*', '[*]')
    .replaceAll('?', '[?]')
    .replaceAll('{', '[{]');
}

/**
 * Resolve an entry `basePath` to a glob-escaped POSIX path relative to
 * `configDirectory`. Absolute base paths are rebased via `path.relative`.
 * Empty / "." means the config root itself (no prefix).
 *
 * A basePath resolving outside `configDirectory` is rejected: the resulting
 * `../`-prefixed (or cross-root) patterns could never match under the Go
 * engine's single-match-root model, so it fails fast instead of silently
 * contributing nothing (mirrors the Go `ResolveBasePaths` rejection).
 */
export function resolveRelativeBasePath(
  basePath: string,
  configDirectory: string,
): string {
  const root = configDirectory || process.cwd();
  const absoluteBase = path.resolve(root, basePath);
  let relative = path.relative(root, absoluteBase).replace(/\\/g, '/');
  if (
    relative === '..' ||
    relative.startsWith('../') ||
    isAbsolutePattern(relative)
  ) {
    throw new Error(
      `basePath "${basePath}" resolves outside the config match root "${root}"; ` +
        'basePath must be a subdirectory of the config root',
    );
  }
  relative = path.posix.normalize(relative);
  if (relative === '.' || relative === '') {
    return '';
  }
  return escapeGlobBasePath(relative).replace(/^\.\//, '');
}

/**
 * Rebase one glob or path pattern under the glob-escaped `relativeBase`
 * produced by {@link resolveRelativeBasePath}.
 *
 * - Leading `!` negation is preserved (odd count → negated).
 * - Leading `./` is stripped before joining.
 * - Absolute patterns are left unchanged (not prefixed).
 * - Empty relativeBase is a no-op besides `./` cleanup.
 */
export function rebasePattern(pattern: string, relativeBase: string): string {
  let negated = false;
  let body = pattern;
  while (body.startsWith('!')) {
    negated = !negated;
    body = body.slice(1);
  }

  body = body.replace(/\\/g, '/');
  while (body.startsWith('./')) {
    body = body.slice(2);
  }

  if (body === '' || isAbsolutePattern(body)) {
    return (negated ? '!' : '') + body;
  }

  if (!relativeBase) {
    return (negated ? '!' : '') + body;
  }

  // path.posix.join collapses `a` + `../b` and keeps `**` segments intact.
  // Strip trailing slashes so the result matches Go's path.Join, which drops
  // them — trailing-slash patterns must not behave differently per side.
  const joined = path.posix.join(relativeBase, body).replace(/\/+$/, '');
  return (negated ? '!' : '') + joined;
}

function rebaseFilesSelector(selector: unknown, relativeBase: string): unknown {
  if (typeof selector === 'string') {
    return rebasePattern(selector, relativeBase);
  }
  if (Array.isArray(selector)) {
    return selector.map((pattern) =>
      typeof pattern === 'string'
        ? rebasePattern(pattern, relativeBase)
        : pattern,
    );
  }
  return selector;
}

function rebaseProjectValue(project: unknown, relativeBase: string): unknown {
  if (typeof project === 'string') {
    return rebasePattern(project, relativeBase);
  }
  if (Array.isArray(project)) {
    return project.map((item) =>
      typeof item === 'string' ? rebasePattern(item, relativeBase) : item,
    );
  }
  return undefined;
}

/**
 * Desugar one config entry's `basePath` into ordinary relative patterns and
 * drop the `basePath` field. Entries without `basePath` are returned as-is
 * (still without a residual `basePath` key).
 *
 * When `basePath` is set and `files` is omitted but the entry still carries
 * rules / plugins / languageOptions / settings, a catch-all
 * `files: ["<base>/**"]` is injected so the entry is scoped to that subtree —
 * matching ESLint's "bare basePath applies under that directory" intent.
 */
export function applyBasePathToEntry(
  entry: Record<string, unknown>,
  configDirectory: string,
): Record<string, unknown> {
  const hasBasePath = Object.prototype.hasOwnProperty.call(entry, 'basePath');
  if (!hasBasePath) {
    return entry;
  }

  const rawBasePath = entry.basePath;
  if (typeof rawBasePath !== 'string') {
    // Caller validates; defensive no-op strip.
    const { basePath: _dropped, ...rest } = entry;
    return rest;
  }

  const relativeBase = resolveRelativeBasePath(rawBasePath, configDirectory);
  const { basePath: _dropped, ...rest } = entry;
  const next: Record<string, unknown> = { ...rest };

  if (Array.isArray(next.files)) {
    next.files = next.files.map((selector) =>
      rebaseFilesSelector(selector, relativeBase),
    );
  }

  if (Array.isArray(next.ignores)) {
    next.ignores = next.ignores.map((pattern) =>
      typeof pattern === 'string'
        ? rebasePattern(pattern, relativeBase)
        : pattern,
    );
  }

  if (isRecord(next.languageOptions)) {
    const languageOptions = { ...next.languageOptions };
    if (isRecord(languageOptions.parserOptions)) {
      const parserOptions = { ...languageOptions.parserOptions };
      if (Object.prototype.hasOwnProperty.call(parserOptions, 'project')) {
        const rebased = rebaseProjectValue(parserOptions.project, relativeBase);
        if (rebased !== undefined) {
          parserOptions.project = rebased;
        }
      }
      languageOptions.parserOptions = parserOptions;
    }
    next.languageOptions = languageOptions;
  }

  const hasFiles = Object.prototype.hasOwnProperty.call(next, 'files');
  const scopesToSubtree =
    next.rules !== undefined ||
    next.plugins !== undefined ||
    next.languageOptions !== undefined ||
    next.settings !== undefined;

  // Scoped entry with basePath but no files → limit to the base subtree.
  if (!hasFiles && scopesToSubtree) {
    const catchAll = relativeBase ? path.posix.join(relativeBase, '**') : '**';
    next.files = [catchAll];
  }

  return next;
}

File: internal/config/base_path.go
package config

import (
	"fmt"
	"path"
	"strings"

	"github.com/microsoft/typescript-go/shim/tspath"
	"github.com/microsoft/typescript-go/shim/vfs"
)

// ResolveBasePaths desugars per-entry BasePath into ordinary relative
// files / ignores / parserOptions.project patterns, then clears BasePath.
//
// configDirectory is the match root the rest of the engine already uses:
// config-file directory for auto-discovered configs, cwd for --config.
// Relative BasePath values resolve against it. Absolute BasePath values are
// rebased via ConvertToRelativePath. fsys supplies filesystem case sensitivity
// (as the rest of config matching does) so an absolute basePath whose casing
// differs from the match root still resolves on case-insensitive filesystems.
//
// A BasePath that resolves outside configDirectory is rejected: its desugared
// patterns would be ../-prefixed and could never match under the engine's
// single-match-root model, indistinguishable from a typo. ESLint reports such
// files as "ignored because outside of base path"; rslint fails fast instead
// so the misconfiguration is actionable.
//
// JS/TS configs normally desugar basePath in Node before the payload reaches
// Go; this path covers JSON/JSONC configs and any raw ConfigEntry constructed
// with BasePath still set.
func ResolveBasePaths(config RslintConfig, configDirectory string, fsys vfs.FS) (RslintConfig, error) {
	if len(config) == 0 {
		return config, nil
	}
	useCaseSensitive := fsys == nil || fsys.UseCaseSensitiveFileNames()
	for i := range config {
		if err := resolveBasePathForEntry(&config[i], configDirectory, useCaseSensitive); err != nil {
			return nil, fmt.Errorf("config entry at index %d: %w", i, err)
		}
	}
	return config, nil
}

func resolveBasePathForEntry(entry *ConfigEntry, configDirectory string, useCaseSensitive bool) error {
	if entry.BasePath == "" {
		return nil
	}

	relativeBase, err := resolveRelativeBasePath(entry.BasePath, configDirectory, useCaseSensitive)
	if err != nil {
		return err
	}
	entry.BasePath = ""

	if len(entry.Files) > 0 {
		rebases := make([]string, len(entry.Files))
		for i, pattern := range entry.Files {
			rebases[i] = rebasePattern(pattern, relativeBase)
		}
		entry.Files = rebases
	}
	if len(entry.FilePatternGroups) > 0 {
		groups := make([][]string, len(entry.FilePatternGroups))
		for i, group := range entry.FilePatternGroups {
			rebases := make([]string, len(group))
			for j, pattern := range group {
				rebases[j] = rebasePattern(pattern, relativeBase)
			}
			groups[i] = rebases
		}
		entry.FilePatternGroups = groups
	}
	if len(entry.Ignores) > 0 {
		rebases := make([]string, len(entry.Ignores))
		for i, pattern := range entry.Ignores {
			rebases[i] = rebasePattern(pattern, relativeBase)
		}
		entry.Ignores = rebases
	}
	if entry.LanguageOptions != nil && entry.LanguageOptions.ParserOptions != nil {
		projects := entry.LanguageOptions.ParserOptions.Project
		if len(projects) > 0 {
			rebases := make([]string, len(projects))
			for i, project := range projects {
				rebases[i] = rebasePattern(project, relativeBase)
			}
			entry.LanguageOptions.ParserOptions.Project = rebases
		}
	}

	// Scoped entry with basePath but no files → limit to the base subtree,
	// matching ESLint's "bare basePath applies under that directory" intent.
	if !hasFileSelectors(*entry) && entryScopesToSubtree(*entry) {
		catchAll := "**"
		if relativeBase != "" {
			catchAll = path.Join(relativeBase, "**")
		}
		entry.Files = []string{catchAll}
	}
	return nil
}

func entryScopesToSubtree(entry ConfigEntry) bool {
	return entry.Rules != nil ||
		entry.Plugins != nil ||
		entry.Settings != nil ||
		entry.LanguageOptions != nil
}

// resolveRelativeBasePath returns a glob-escaped POSIX-ish path relative to
// configDirectory. Empty means the config root itself (no prefix). It rejects
// a basePath that resolves outside configDirectory, since the resulting
// ../-prefixed patterns could never match under the engine's single-match-root
// model.
func resolveRelativeBasePath(basePath string, configDirectory string, useCaseSensitive bool) (string, error) {
	if basePath == "" {
		return "", nil
	}
	// tspath.ResolvePath normalizes to forward slashes, but callers (and test
	// fixtures) may hand in a configDirectory with native separators. Normalize
	// once so the containment check and relative rebase compare like for like.
	if configDirectory != "" {
		configDirectory = tspath.NormalizePath(configDirectory)
	}
	absolute := tspath.ResolvePath(configDirectory, basePath)
	if _, ok := RelativePathWithinConfigRoot(absolute, configDirectory, useCaseSensitive); !ok {
		return "", fmt.Errorf(
			"key \"basePath\": value %q resolves outside the config match root %q; basePath must be a subdirectory of the config root",
			basePath,
			configDirectory,
		)
	}
	relative := tspath.ConvertToRelativePath(absolute, tspath.ComparePathsOptions{
		UseCaseSensitiveFileNames: useCaseSensitive,
		CurrentDirectory:          configDirectory,
	})
	relative = strings.ReplaceAll(tspath.NormalizePath(relative), "\\", "/")
	relative = strings.TrimPrefix(relative, "./")
	if relative == "." || relative == "" {
		return "", nil
	}
	return escapeGlobBasePath(relative), nil
}

// escapeGlobBasePath escapes glob metacharacters in a resolved basePath so the
// directory is treated literally when spliced into glob patterns. ESLint treats
// basePath as a literal path; without escaping, a basePath such as
// "packages/[locale]" would be interpreted as a character class and never match.
//
// Escaping uses character classes ([*], [?], [[] ...), which the doublestar and
// picomatch matchers understand on every platform and which survive the
// tspath.NormalizePath applied to patterns before matching. A bare ']' or '}'
// outside a class/brace pair is already literal to the matchers, so only the
// characters that open a class/brace or introduce a wildcard need escaping. A
// leading '!' (basePath "!vendor") has no class form the matchers accept as
// literal, so it remains a known limitation.
func escapeGlobBasePath(s string) string {
	return strings.NewReplacer(
		"[", "[[]",
		"*", "[*]",
		"?", "[?]",
		"{", "[{]",
	).Replace(s)
}

// rebasePattern prefixes a relative glob/path with the glob-escaped
// relativeBase produced by resolveRelativeBasePath.
// Leading ! negation is preserved; absolute patterns are left unchanged;
// empty relativeBase only strips a leading ./.
func rebasePattern(pattern string, relativeBase string) string {
	negated := false
	body := pattern
	for strings.HasPrefix(body, "!") {
		negated = !negated
		body = strings.TrimPrefix(body, "!")
	}
	body = strings.ReplaceAll(body, "\\", "/")
	for strings.HasPrefix(body, "./") {
		body = strings.TrimPrefix(body, "./")
	}
	if body == "" || isAbsolutePattern(body) {
		if negated {
			return "!" + body
		}
		return body
	}
	if relativeBase == "" {
		if negated {
			return "!" + body
		}
		return body
	}
	joined := path.Join(relativeBase, body)
	if negated {
		return "!" + joined
	}
	return joined
}

// isAbsolutePattern reports whether pattern is an absolute path in any form
// tspath recognizes (POSIX, UNC, Windows drive-letter).
func isAbsolutePattern(pattern string) bool {
	return tspath.GetEncodedRootLength(pattern) > 0
}
